geonode/harvesting/harvesters/wms.py

```diff
@@ -104,66 +136,179 @@ def check_availability(self, timeout_seconds: typing.Optional[int] = 5) -> bool:
+            element.nsmap)
+        service = get_xpath_elements(element, "wms:Service", nsmap)[0]
+        contact = get_xpath_elements(service, "wms:ContactInformation", nsmap)[0]
+        contact_person = get_xpath_elements(contact, "wms:ContactPersonPrimary", nsmap)[0]
+        address = get_xpath_elements(contact, "wms:ContactAddress", nsmap)[0]
+        return {
+            'role': '',
+            'name': get_xpath_value(contact_person, "wms:ContactPerson", nsmap),
+            'organization': get_xpath_value(contact_person, "wms:ContactOrganization", nsmap),
+            'position': get_xpath_value(contact, "wms:ContactPosition", nsmap),
+            'phone_voice': get_xpath_value(contact_person, "wms:ContactVoiceTelephone", nsmap),
+            'phone_facsimile': get_xpath_value(contact_person, "wms:ContactFacsimileTelephone", nsmap),
+            'address_delivery_point': '',
+            'address_city': get_xpath_value(address, "wms:City", nsmap),
+            'address_administrative_area': get_xpath_value(address, "wms:StateOrProvince", nsmap),
+            'address_postal_code': get_xpath_value(address, "wms:PostCode", nsmap),
+            'address_country': get_xpath_value(address, "wms:Country", nsmap),
+            'address_email': get_xpath_value(contact, "wms:ContactElectronicMailAddress", nsmap),
+        }
 
-    def _get_useful_layers(self) -> typing.List[etree.Element]:
-        get_capabilities_response = self.http_session.get(
-            self.remote_url,
-            params={
-                "service": "WMS",
-                "version": "1.3.0",
-                "request": "GetCapabilities",
-            }
-        )
-        get_capabilities_response.raise_for_status()
-        root = etree.fromstring(get_capabilities_response.content, parser=XML_PARSER)
-        nsmap = _get_nsmap(root.nsmap)
-        useful_layers_elements = []
-        leaf_layers = root.xpath("//wms:Layer[not(.//wms:Layer)]", namespaces=nsmap)
-        for layer_element in leaf_layers:
-            title = layer_element.xpath("wms:Title/text()", namespaces=nsmap)[0]
-            if self.layer_title_filter is not None:
-                if self.layer_title_filter.lower() not in title.lower():
-                    continue
-            useful_layers_elements.append(layer_element)
-        return useful_layers_elements
+    def _layer_element_to_json(self, layer_element: etree.Element) -> dict:
+        """Return json of layer from xml element"""
+        nsmap = _get_nsmap(
+            layer_element.nsmap)
+        name = get_xpath_value(
+            layer_element, "wms:Name", nsmap)
+        title = get_xpath_value(
+            layer_element, "wms:Title", nsmap)
+        abstract = get_xpath_value(
+            layer_element, "wms:Abstract", nsmap)
+        try:
+            keywords = get_xpath_elements(
+                get_xpath_elements(layer_element, "wms:KeywordList", nsmap)[0],
+                "wms:Keyword/text()", nsmap
+            )
```

Comment:
This can be made simpler by relying on XPath expressions a bit more (their indexing starts at 1 instead of zero though):

```python
keywords = layer_element.xpath("wms:KeywordList[1]/wms:Keyword/text()", namespaces=nsmap)
```